runtime/lua/vim/semver.lua

```diff
@@ -0,0 +1,262 @@
+  return nil
+end
+
+---@private
+--- Extracts the major, minor, patch and preprelease and build components from
+--- `version`.
+---@param version string Version string
+local function extract_components_strict(version)
+  local major, minor, patch, prerelease_and_build = version:match('^v?(%d+)%.(%d+)%.(%d+)(.*)$')
+  return tonumber(major), tonumber(minor), tonumber(patch), prerelease_and_build
+end
+
+---@private
+--- Extracts the major, minor, patch and preprelease and build components from
+--- `version`. When `minor` and `patch` components are not found (nil), coerce
+--- them to 0.
+---@param version string Version string
+local function extract_components_loose(version)
+  local major, minor, patch, prerelease_and_build = version:match('^v?(%d+)%.?(%d*)%.?(%d*)(.*)$')
+  major = tonumber(major)
+  minor = tonumber(minor) or 0
+  patch = tonumber(patch) or 0
+  return major, minor, patch, prerelease_and_build
+end
+
+---@private
+--- Validates the prerelease and build string e.g. "-rc1+build.0". If the
+--- prerelease, build or both are valid forms then it will return true, if it
+--- is not of any valid form, it will return false.
+---@param prerelease_and_build string
+---@return boolean
+local function is_prerelease_and_build_valid(prerelease_and_build)
+  local has_build = prerelease_and_build:match('^%+[%w%.]+$')
+  local has_prerelease = prerelease_and_build:match('^%-[%w%.]+$')
+  local has_prerelease_and_build = prerelease_and_build:match('^%-[%w%.]+%+[%w%.]+')
+  return has_build or has_prerelease or has_prerelease_and_build
+end
+
+---@private
+---@param prerelease_and_build string
+---@return boolean
+local function has_prerelease_and_build(prerelease_and_build)
+  return prerelease_and_build ~= nil and prerelease_and_build ~= ''
+end
+
+---@private
+---@param version string
+---@return string
+local function create_err_msg(version)
+  return string.format('%s is not a valid version', version)
+end
+
+--- Parses a semantically formatted version string into a table.
+---
+--- Supports leading "v" and leading and trailing whitespace in the version
+--- string, that means `"v1.0.1-rc1+build.2"`, `"1.0.1-rc1+build.2"`, `" v1.0.1-rc1+build.2"`
+--- and `"v1.0.1-rc1+build.2 "` will all produce the table:
+---
+--- `{ major = 1, minor = 0, patch = 1, prerelease = 'rc1', build = 'build.2' }`
+---
+---@param version string Version string formatted semantically.
+---@param opts table|nil Options to configure parsing behaviour which takes the following optional keys:
+---                      - strict (boolean):  when set to `true` an error will be thrown for version
+---                      strings which are not conforming to semver sepecifications, this means that
+---                      `semver.parse('v1.2)` will throw an error. When set to `false`,
+---                      `semver.parse('v1.2)` will coerce 'v1.2' to 'v1.2.0' and return the table:
+---                      `{ major = 1, minor = 2, patch = 0 }`. Defaults to false.
```

Comment:
```suggestion
---                      - strict (boolean): if `true`, raise an error if the version string does not conform to the specification; if `false` (default), parse missing fields as `0` (e.g., `semver.parse('v1.2)` will return `{ major = 1, minor = 2, patch = 0 }`.`
```

could (should) add a link to the specification?